compiler/rustc_middle/src/ty/fold.rs

```diff
@@ -876,6 +876,52 @@ impl<'tcx> PredicateVisitor<'tcx> for HasEscapingVarsVisitor {
     }
 }
 
+crate struct CountBoundVars {
+    crate outer_index: ty::DebruijnIndex,
+    crate bound_tys: FxHashSet<ty::BoundTy>,
+    crate bound_regions: FxHashSet<ty::BoundRegion>,
+    crate bound_consts: FxHashSet<ty::BoundVar>,
+}
+
+impl<'tcx> TypeVisitor<'tcx> for CountBoundVars {
+    fn visit_binder<T: TypeFoldable<'tcx>>(&mut self, t: &Binder<T>) -> bool {
+        self.outer_index.shift_in(1);
+        let result = t.super_visit_with(self);
+        self.outer_index.shift_out(1);
+        result
+    }
+
+    fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+        match t.kind {
```

Comment:
can you exit early with `ty.has_vars_bound_at_or_above(self.outer_index)` here? Should greatly reduce the perf impact I think

Reply:
you can also implement `PredicateVisitor` here to do the same for `ty::Predicate`